src/zenml/utils/deprecation_utils.py

```diff
@@ -0,0 +1,139 @@
+
+    ```python
+    from pydantic import BaseModel
+    from typing import Optional
+
+    class MyModel(BaseModel):
+        deprecated: Optional[int] = None
+
+        old_name: Optional[str] = None
+        new_name: str
+
+        _deprecation_validator = deprecate_pydantic_attributes(
+            "deprecated", ("old_name", "new_name")
+        )
+    ```
+
+    Args:
+        *attributes: List of attributes to deprecate. This is either the name
+            of the attribute to deprecate, or a tuple containing the name of
+            the deprecated attribute and it's replacement.
+
+    Returns:
+        Pydantic validator class method to be used on BaseModel subclasses
+        to deprecate or migrate attributes.
+    """
+
+    @root_validator(pre=True, allow_reuse=True)
+    def _deprecation_validator(
+        cls: Type[BaseModel], values: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """Pydantic validator function for deprecating pydantic attributes.
+
+        Args:
+            cls: The class on which the attributes are defined.
+            values: All values passed at model initialization.
+
+        Raises:
+            AssertionError: If either the deprecated or replacement attribute don't
+                exist.
+            TypeError: If the deprecated attribute is a required attribute.
+            ValueError: If the deprecated attribute and replacement attribute
+                contain different values.
+
+        Returns:
+            Input values with potentially migrated values.
+        """
+        for attribute in attributes:
+            if isinstance(attribute, str):
+                deprecated_attribute = attribute
+                replacement_attribute = None
+            else:
+                deprecated_attribute, replacement_attribute = attribute
+
+                assert (
+                    replacement_attribute in cls.__fields__
+                ), f"Unable to find attribute {replacement_attribute}."
+
+            assert (
+                deprecated_attribute in cls.__fields__
+            ), f"Unable to find attribute {deprecated_attribute}."
+
+            if cls.__fields__[deprecated_attribute].required:
+                raise TypeError(
+                    f"Unable to deprecate attribute '{deprecated_attribute}' "
+                    f"of class {cls.__name__}. In order to deprecate an "
+                    "attribute, it needs to be a non-required attribute. "
+                    "To do so, mark the attribute with an `Optional[...] type "
+                    "annotation."
+                )
+
+            if deprecated_attribute in values:
+                if replacement_attribute is None:
+                    warnings.warn(
+                        f"The attribute {deprecated_attribute} of class "
+                        f"{cls.__name__} will be deprecated soon.",
+                        DeprecationWarning,
+                    )
```

Comment:
Mhm is it something worthy of a warning though? In case of a replacement attribute which ZenML can automatically handle the migration for, I'm not sure it's necessary to constantly warn them each time?

Reply:
will we handle this migration long-term? If so, then I agree that it's unnecessary. I was assuming this deprecation util would only be used to migrate old attributes for a few versions and would then be taken out again, in which case passing the old attribute might lead to errors at some point.

Reply:
The current implementation already removes the value of the deprecated attribute and sets it on the new one. Would that cover long-term migration? I can see two potential issues:
* We only migrate it for stack components that get instantiated. So if someone doesn't use an orchestrator for half a year, we won't migrate it. At the same time, we wouldn't print a deprecation warning either so I think this is probably not that big of an issue?
* I'm not sure if the changes currently get written to the ZenStore once updated, but I'll make sure they are.